### PR TITLE
[FLINK-1799][scala] Fix handling of generic arrays

### PR DESCRIPTION
**Pull request by aljoscha**

### CODE REVIEW

**Comment by StephanEwen** [comment]
Instead of the switch/case statement, can't you use `PrimitiveArrayTypeInfo.getInfoFor(Class<X> type)`? Seems simpler.

Also, this is a good change to be backed by a Unit test case, rather than an Integration test case.

I really think that such changes should be guarded by more fine grained tests (Unit Tests(, rather than the "big hammer" kind of test (have a distributed program that uses this). Out testing time goes through the roof otherwise.

I fund that unit tests are a cycle or two more work to write, but typically also get you thinking better about what cases the tests should cover.

**Comment by aljoscha** [comment]
Using ``PrimitiveArrayTypeInfo.getInfoFor(Class<X> type)`` is not easily possible since that requires an array type, which I don't have available at this point.

For the tests you are absolutely correct, I will change them.

**Comment by aljoscha** [comment]
I removed the IT case and added a test instead.

**Comment by aljoscha** [comment]
Any more comments?

**Comment by tillrohrmann** [comment]
Maybe we could add in the test case the case where we have an ```Object``` array. Other than that LGTM.

**Comment by aljoscha** [comment]
I added the test case @tillrohrmann suggested. Also helped me find a bug. :relieved:

**Comment by aljoscha** [comment]
Any more thoughts? Otherwise I would like to merge this.

File: flink-java/src/main/java/org/apache/flink/api/java/typeutils/ObjectArrayTypeInfo.java [hunk 1]
```diff
@@ -143,6 +143,18 @@ else if (type instanceof Class<?> && ((Class<?>) type).isArray()
...
+	/**
+	 * Creates a new {@link org.apache.flink.api.java.typeutils.ObjectArrayTypeInfo} from a
+	 * {@link TypeInformation} for the component type.
+	 *
+	 * <p>
+	 * This must be used in cases where the complete type of the array is not available as a
+	 * {@link java.lang.reflect.Type} or {@link java.lang.Class}.
+	 */
+	public static <T, C> ObjectArrayTypeInfo<T, C> getInfoFor(TypeInformation<C> componentInfo) {
+		return new ObjectArrayTypeInfo<T, C>(Object[].class, componentInfo.getTypeClass(), componentInfo);
```
Comment (StephanEwen) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/ObjectArrayTypeInfo.java]: Is `Object[].class` actually correct? Why not give the class of the real array type? There are a few ways to get that, for example `Array.newInstance(clazz, 0).getClass();`

Comment (aljoscha) [inline flink-java/src/main/java/org/apache/flink/api/java/typeutils/ObjectArrayTypeInfo.java]: That's right, I didn't think of that. Will change.

**Comment by aljoscha** [comment]
I fixed @StephanEwen's complaint. It was incorrect but the type class of TypeInformation does not seem to be used in any places where it matters.

**Comment by aljoscha** [comment]
So, any thoughts about merging this?

**Comment by StephanEwen** [comment]
Yo, looks fine now

**Pull request merged**